devices/src/virtio/vsock/csm/connection.rs

```diff
@@ -0,0 +1,632 @@
+                    // data.
+                    self.state = ConnState::LocalClosed;
+                    self.expiry = Some(
+                        Instant::now() + Duration::from_millis(defs::CONN_SHUTDOWN_TIMEOUT_MS),
+                    );
+                    pkt.set_op(uapi::VSOCK_OP_SHUTDOWN)
+                        .set_flag(uapi::VSOCK_FLAGS_SHUTDOWN_RCV)
+                        .set_flag(uapi::VSOCK_FLAGS_SHUTDOWN_SEND);
+                } else {
+                    // On a successful data read, we fill in the packet with the RW op, and length of
+                    // the read data.
+                    pkt.set_op(uapi::VSOCK_OP_RW).set_len(read_cnt as u32);
+                }
+            }
+            Err(e) => {
+                if e.kind() == std::io::ErrorKind::WouldBlock {
+                    // If reading from the host stream would've blocked now, it might succeed later.
+                    return Err(VsockError::NoData);
+                }
+                // We don't know how to handle any other I/O error, so it's time to kill this
+                // connection.
+                pkt.set_op(uapi::VSOCK_OP_RST);
+            }
+        };
+
+        self.rx_cnt += Wrapping(pkt.len());
+        self.last_fwd_cnt_to_peer = self.fwd_cnt;
+
+        Ok(())
+    }
+
+    /// Deliver a guest-generated packet to this connection.
+    ///
+    /// This forwards the data in RW packets to the host stream, and absorbs control packets,
+    /// using them to manage the internal connection state.
+    ///
+    /// Returns:
+    /// always `Ok(())`: the packet has been consumed;
+    ///
+    fn send_pkt(&mut self, pkt: &VsockPacket) -> VsockResult<()> {
+        // Update the peer credit information.
+        self.peer_buf_alloc = pkt.buf_alloc();
+        self.peer_fwd_cnt = Wrapping(pkt.fwd_cnt());
+
+        match self.state {
+            // Most frequent case: this is an established connection that needs to forward some data
+            // to the host stream. Also works for a connection that has begun shutting down, but the
+            // peer still has some data to send.
+            ConnState::Established | ConnState::PeerClosed(_, false)
+                if pkt.op() == uapi::VSOCK_OP_RW =>
+            {
+                if pkt.buf().is_none() {
+                    info!(
+                        "vsock: dropping empty data packet from guest (lp={}, pp={}",
+                        self.local_port, self.peer_port
+                    );
+                    return Ok(());
+                }
+
+                // Unwrapping here is safe, since we just checked `pkt.buf()` above.
+                let buf_slice = &pkt.buf().unwrap()[..(pkt.len() as usize)];
+                if let Err(err) = self.send_bytes(buf_slice) {
+                    // If we can't write to the host stream, that's an unrecoverable error, so we'll
+                    // terminate this connection.
+                    info!(
```

Comment:
Shouldn t this be an error since we are terminating the vsock connection at an unexpected time?